sky/clouds/service_catalog/gcp_catalog.py

```diff
@@ -373,3 +406,87 @@ def get_region_zones_for_accelerators(
     """Returns a list of regions for a given accelerators."""
     df = _get_accelerator(_df, accelerator, count, region=None)
     return common.get_region_zones(df, use_spot)
+
+
+def check_host_accelerator_compatibility(instance_type: str,
+                                         accelerators: Optional[Dict[str, int]],
+                                         zone: Optional[str] = None) -> None:
+    """Check if the instance type is compatible with the accelerators."""
+    if accelerators is None:
+        if instance_type.startswith('a2-highgpu-'):
+            # NOTE: While it is allowed to use A2 machines as CPU-only nodes,
+            # we exclude this case as it is uncommon and undesirable.
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'A2 instance types should be used with A100 GPUs. '
+                    'Either use other instance types or specify the '
+                    'accelerators as A100.')
+        return
+
+    acc = list(accelerators.items())
+    assert len(acc) == 1, acc
+    acc_name, acc_count = acc[0]
+
+    if 'tpu' in acc_name.lower():
+        # TODO(woosuk): Add validation for TPUs.
+        return
+    acc_name = acc_name.upper()
+
+    # Treat A100 as a special case.
+    if acc_name == 'A100':
+        # A100 must be attached to A2 instance type.
+        if not instance_type.startswith('a2-highgpu-'):
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'A100 GPUs cannot be attached to {instance_type}. '
+                    'Use A2 instance type instead.')
+        a100_instance_type = _A100_INSTANCE_TYPES[acc_count]
+        if instance_type != a100_instance_type:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'A100:{acc_count} cannot be attached to {instance_type}. '
+                    f'Use {a100_instance_type} instead.')
+        return
```

Comment:
Testing this on master:
```
» sky gpunode --gpus A100 -t a2-highgpu-1g
I 07-29 17:26:11 optimizer.py:605] == Optimizer ==
I 07-29 17:26:11 optimizer.py:628] Estimated cost: $3.7 / hour
I 07-29 17:26:11 optimizer.py:628]
I 07-29 17:26:11 optimizer.py:675] Considered resources (1 node):
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]  CLOUD   INSTANCE        ACCELERATORS   COST ($)   CHOSEN
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]  GCP     a2-highgpu-1g   A100:1         3.67          ✔
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]
Launching a new cluster 'sky-gpunode-zongheng'. Proceed? [Y/n]: n
Aborted!

» sky gpunode  -t a2-highgpu-1g                                                       1 ↵
I 07-29 17:26:20 optimizer.py:605] == Optimizer ==
I 07-29 17:26:20 optimizer.py:628] Estimated cost: $0.7 / hour
I 07-29 17:26:20 optimizer.py:628]
I 07-29 17:26:20 optimizer.py:675] Considered resources (1 node):
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]  CLOUD   INSTANCE        ACCELERATORS   COST ($)   CHOSEN
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]  GCP     a2-highgpu-1g   -              0.74          ✔
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]
Launching a new cluster 'sky-gpunode-zongheng'. Proceed? [Y/n]:
```
I thought A100s on GCP are special in that they come with the a2 VMs, and we should not allow the first case (attaching A100 to a2 VM) to happen?

Reply:
If I understand correctly, A100 GPUs must be attached to A2 VMs, just like other types of GPUs must be attached to N1 VMs. TPU nodes are the only exceptions that host VMs come with the accelerators.

Reply:
I see. On master I could launch a no-GPU version of `a2-highgpu-1g`. On GCP console, there doesn't appear a way to do so -- one must select A100 GPUs first which will unlock the choice of an `a2` VM.  I think this PR's treatment of banning CPU-only A2 is good as it's consistent with the GCP console (not the API loophole).